### PR TITLE
Calculate Nav Source and fix popping params

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

- Fix ShellNavigationSource so it's correctly calculated based on diff of source/dest
- When popping using ".." don't clear parameters on previous page 

### Platforms Affected ### 
- Core/XAML (all platforms)

### Testing Procedure ###
- unit tests pass
- UI tests pass

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 1]
```diff
@@ -1653,6 +1653,47 @@ bool ProposeNavigation(ShellNavigationSource source, ShellNavigationState propos
...
+
+			if(targetPaths.Length < currentPaths.Length)
+			{
+				if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
+					return ShellNavigationSource.Remove;
+
+				return ShellNavigationSource.Pop;
+			}
+
+			if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
```
Comment (pictos) [inline Xamarin.Forms.Core/Shell/Shell.cs:1694]: On my first suggestion I added two locals variables
```
var targetPathsLength = targetPaths.Length;
var currentPathsLength = currentPaths.Length;
```
so without them, the code will fail.

```suggestion
			if (targetPaths[targetPaths.Length - 1] == currentPaths[currentPaths.Length - 1])
```

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 2]
```diff
@@ -1653,6 +1652,79 @@ bool ProposeNavigation(ShellNavigationSource source, ShellNavigationState propos
...
+
+					if (i == targetPaths.Length - 1)
+						return ShellNavigationSource.Push;
+				}
+			}
+
+			if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
+				return ShellNavigationSource.Insert;
+
+			return ShellNavigationSource.Push;
```
Comment (pictos) [inline Xamarin.Forms.Core/Shell/Shell.cs:1725]: @PureWeen  just a micro-opt avoiding to call the `get` method inside the props. (:
```suggestion
		var targetPathsLength = targetPaths.Length;
			var currentPathsLength = currentPaths.Length;

			if (targetPathsLength < 4 || currentPathsLength < 4)
				return ShellNavigationSource.Unknown;

			if (targetPaths[1] != currentPaths[1])
				return ShellNavigationSource.ShellItemChanged;
			if (targetPaths[2] != currentPaths[2])
				return ShellNavigationSource.ShellSectionChanged;
			if (targetPaths[3] != currentPaths[3])
				return ShellNavigationSource.ShellContentChanged;
'
			if (targetPathsLength == currentPathsLength)
				return ShellNavigationSource.Unknown;

			if(targetPathsLength < currentPathsLength)
			{
				for (var i = 0; i < targetPathsLength; i++)
				{
					var targetPath = targetPaths[i];
					if (targetPath != currentPaths[i])
						break;

					if (i == targetPathsLength - 1)
					{
						if (targetPathsLength == 4)
							return ShellNavigationSource.PopToRoot;

						return ShellNavigationSource.Pop;
					}
				}

				if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
					return ShellNavigationSource.Remove;

				if (targetPathsLength == 4)
					return ShellNavigationSource.PopToRoot;

				return ShellNavigationSource.Pop;
			}
			else if(targetPathsLength > currentPathsLength)
			{
				for (var i = 0; i < currentPathsLength; i++)
				{
					if (targetPaths[i] != currentPaths[i])
						break;

					if (i == targetPathsLength - 1)
						return ShellNavigationSource.Push;
				}
			}

			if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
				return ShellNavigationSource.Insert;

			return ShellNavigationSource.Push;
```

**Review by pictos — APPROVED** [review]
All works here on my side. I have just a small suggestion but that doesn't invalidate this PR.

File: Xamarin.Forms.Core/Shell/Shell.cs (hunk 2 above)
Comment (PureWeen) [reply Xamarin.Forms.Core/Shell/Shell.cs]: 👍

**Pull request merged**